python/gps/algorithm/algorithm_traj_opt.py

```diff
@@ -98,15 +110,113 @@ def _stepadjust(self, m):
+                                                        np.sum(np.log(np.diag(demo_traj[itr_i].chol_pol_covar[t, :, :])))
+            # Sum over the distributions and time.
+            samples_logiw[i] = logsum(np.sum(samples_logprob[i], 1), 0)
+
+        # Assume only one condition for the samples.
+        assert Md == 1
+        for idx in xrange(Md):
+            if M == 1:
+                i = 0
+            else:
+                i = idx
+            # Evaluate demo prob. under sample distributions.
+            for itr_i in xrange(itr + 1):
+                traj = self.traj_distr[itr_i][i]
+                for j in xrange(demoX[idx].shape[0]):
+                    for t in xrange(self.T - 1):
+                        diff = traj.k[t, :] + \
+                                traj.K[t, :, :].dot(demoX[idx][j, t, :]) - demoU[idx][j, t, :]
+                        demos_logprob[idx][itr_i, t, j] = -0.5 * np.sum(diff * (traj.inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(traj.chol_pol_covar[t, :, :])))
+            # Evaluate demo prob. under demo distributions.
+            for itr_i in xrange(Md):
+                for j in range(demoX[idx].shape[0]):
+                    for t in xrange(self.T - 1):
+                        diff = demo_traj[itr_i].k[t, :] + \
+                                demo_traj[itr_i].K[t, :, :].dot(demoX[idx][j, t, :]) - demoU[idx][j, t, :]
+                        demos_logprob[idx][itr + itr_i + 1, t, j] = -0.5 * np.sum(diff * (demo_traj[itr_i].inv_pol_covar[t, :, :].dot(diff)), 0) - \
+                                                        np.sum(np.log(np.diag(demo_traj[itr_i].chol_pol_covar[t, :, :])))
+            # Sum over the distributions and time.
+            demos_logiw[idx] = logsum(np.sum(demos_logprob[idx], 1), 0)
+
+
+        # Update the learned cost
+        # Transform all the dictionaries to arrays
+        demoU_arr =  np.vstack((self.demo_list.get_U() for i in xrange(self.Md)))
+        demoX_arr =  np.vstack((self.demo_list.get_X() for i in xrange(self.Md)))
+        demoO_arr =  np.vstack((self.demo_list.get_obs() for i in xrange(self.Md)))
+        sampleU_arr = np.vstack((self.sample_list[i].get_U() for i in xrange(M)))
+        sampleX_arr = np.vstack((self.sample_list[i].get_X() for i in xrange(M)))
+        sampleO_arr = np.vstack((self.sample_list[i].get_obs() for i in xrange(M)))
+        demos_logiw = np.vstack((demos_logiw[i] for i in xrange(Md)))
+        samples_logiw = np.vstack((samples_logiw[i] for i in xrange(M)))
+        cost_ioc_quad = self.cost # set the type of cost to be cost_ioc_quad here.
+        self.learned_cost = cost_ioc_quad.update(demoU_arr, demoX_arr, demoO_arr, demos_logiw, sampleU_arr, sampleX_arr, \
+                                                sampleO_arr, samples_logiw, itr, self)
+        
```

Comment:
You can start testing now (just added only 2 minor comments). When you test, I would try to just get past the update call above.

To test it, you will need to set the cost object in hyperparams to be CostIOCQuad (and specify where the demos are coming from). In the long term, we will have two cost objects, one for the ground truth cost and one for the learned cost, but for now, you can just assume one for testing.